tests/test_tangentspace.py

```diff
@@ -1,94 +1,110 @@
-
-
-def test_TangentSpace_inversetransform():
-    """Test inverse transform of Tangent Space."""
-    covset = generate_cov(10, 3)
-    ts = TangentSpace(metric='riemann')
-    ts.fit(covset)
-    t = ts.transform(covset)
-    cov = ts.inverse_transform(t)
-    assert_array_almost_equal(covset, cov)
-
-
-def test_TangentSpace_inversetransform_without_fit():
-    """Test inverse transform of Tangent Space without fit."""
-    covset = generate_cov(10, 3)
-    ts = TangentSpace(metric='identity')
-    tsv = ts.fit_transform(covset)
-    ts = TangentSpace(metric='riemann')
-    cov = ts.inverse_transform(tsv)
-    assert_array_almost_equal(covset, cov)
-
-
-def test_FGDA_init():
-    """Test init of FGDA."""
-    FGDA(metric='riemann')
-
-
-def test_FGDA_fit():
-    """Test Fit of FGDA."""
-    covset = generate_cov(10, 3)
-    labels = np.array([0, 1]).repeat(5)
-    ts = FGDA(metric='riemann')
-    ts.fit(covset, labels)
-
-
-def test_FGDA_transform():
-    """Test transform of FGDA."""
-    covset = generate_cov(10, 3)
-    labels = np.array([0, 1]).repeat(5)
-    ts = FGDA(metric='riemann')
-    ts.fit(covset, labels)
-    ts.transform(covset)
+import pytest
+from pytest import approx
+
+
+@pytest.mark.parametrize("tspace", [TangentSpace, FGDA])
+class TangentSpaceTestCase:
+    def test_tangentspace(self, tspace, get_covmats):
+        n_classes, n_trials, n_channels = 2, 6, 3
+        covmats = get_covmats(n_trials, n_channels)
+        labels = np.array([0, 1]).repeat(n_trials // n_classes)
+        self.clf_transform(tspace, covmats, labels)
+        self.clf_fit_transform(tspace, covmats, labels)
+        self.clf_fit_transform_independence(tspace, covmats, labels)
+        if tspace is TangentSpace:
+            self.clf_transform_wo_fit(tspace, covmats)
+            self.clf_inversetransform(tspace, covmats)
+
+
+class TestTangentSpace(TangentSpaceTestCase):
+    def clf_transform(self, tspace, covmats, labels):
+        n_trials, n_channels, n_channels = covmats.shape
+        ts = tspace().fit(covmats, labels)
+        Xtr = ts.transform(covmats)
+        if tspace is TangentSpace:
+            n_ts = (n_channels * (n_channels + 1)) // 2
+            assert Xtr.shape == (n_trials, n_ts)
+        else:
+            assert Xtr.shape == (n_trials, n_channels, n_channels)
+
+    def clf_fit_transform(self, tspace, covmats, labels):
+        n_trials, n_channels, n_channels = covmats.shape
+        ts = tspace()
+        Xtr = ts.fit_transform(covmats, labels)
+        if tspace is TangentSpace:
+            n_ts = (n_channels * (n_channels + 1)) // 2
+            assert Xtr.shape == (n_trials, n_ts)
+        else:
+            assert Xtr.shape == (n_trials, n_channels, n_channels)
+
+    def clf_fit_transform_independence(self, tspace, covmats, labels):
+        n_trials, n_channels, n_channels = covmats.shape
+        ts = tspace()
+        ts.fit(covmats, labels)
+        # retraining with different size should erase previous fit
```

Comment:
In fact, all algos should have this _re-fit test_.

Reply:
done